tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs

```diff
@@ -0,0 +1,280 @@
+                }
+            }
+        });
+    }
+
+    private void AddExternalToolToContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        menuFlyout.Items.Add(CreateContextMenuItemForTool(tool));
+    }
+
+    private void RemoveExternalToolFromContextMenu(MenuFlyout menuFlyout, ExternalTool tool)
+    {
+        foreach (var menuItem in menuFlyout.Items)
+        {
+            if (menuItem.Tag == tool)
+            {
+                tool.PropertyChanged -= ExternalToolItem_PropertyChanged;
+                menuFlyout.Items.Remove(menuItem);
+                break;
+            }
+        }
+    }
+
+    // This creates a Menu Flyout item for an external tool. It also creates a sub-menu item for pinning/unpinning
+    // and unregistering the tool.
+    private MenuFlyoutItem CreateContextMenuItemForTool(ExternalTool tool)
+    {
+        var imageIcon = new ImageIcon
+        {
+            Source = tool.ToolIcon,
+        };
+
+        var menuItem = new MenuFlyoutItem
+        {
+            Text = tool.Name,
+            Tag = tool,
+            Icon = imageIcon,
+        };
+        menuItem.Click += ExternalToolMenuItem_Click;
+
+        var pinMenuSubItemItem = new MenuFlyoutItem
+        {
+            Text = _pinMenuItemText,
+            Icon = GetFontIcon(CommonHelper.PinGlyph),
+            Tag = tool,
+        };
+        pinMenuSubItemItem.Click += ExternalToolPinUnpin_Click;
+
+        var unPinMenuSubItemItem = new MenuFlyoutItem
+        {
+            Text = _unpinMenuItemText,
+            Icon = GetFontIcon(CommonHelper.UnpinGlyph),
+            Tag = tool,
+        };
+        unPinMenuSubItemItem.Click += ExternalToolPinUnpin_Click;
+
+        var unRegisterMenuSubItemItem = new MenuFlyoutItem
+        {
+            Text = _unregisterMenuItemText,
+            Icon = GetFontIcon(_UnregisterButtonText),
+            Tag = tool,
+        };
+        unRegisterMenuSubItemItem.Click += UnregisterMenuItem_Click;
+
+        var menuSubItemFlyout = new MenuFlyout();
+
+        menuSubItemFlyout.Items.Add(pinMenuSubItemItem);
+        menuSubItemFlyout.Items.Add(unPinMenuSubItemItem);
+        menuSubItemFlyout.Items.Add(unRegisterMenuSubItemItem);
+
+        if (tool.IsPinned)
+        {
+            pinMenuSubItemItem.Visibility = Visibility.Collapsed;
+        }
+        else
+        {
+            unPinMenuSubItemItem.Visibility = Visibility.Collapsed;
+        }
+
+        menuItem.ContextFlyout = menuSubItemFlyout;
+
+        return menuItem;
+    }
+
+    private void ExternalToolItem_PropertyChanged(object? sender, PropertyChangedEventArgs e)
+    {
+        if (sender is not ExternalTool tool)
+        {
+            Debug.Assert(false, "What is this?");
+            return;
+        }
+
+        _dispatcher.TryEnqueue(() =>
+        {
+            MenuFlyout? flyout = this.ContextFlyout as MenuFlyout;
+            Debug.Assert(flyout is not null, "Why does this button not have a menu flyout?");
```

Comment:
If the ContextMenu is needs to be a MenuFlyout, please move this check to the top so this check can be performed earlier and future readers can notice this faster.

Reply:
I wanted to perform the check on the UI thread, not only to avoid race conditions, the just the general goodness of accessing UI elements on the UI thread.